File: frontend/src/services/deployment.ts
import api from '../config/api';

export interface DeploymentConfig {
  id: number;
  name: string;
  application_name: string;
  description?: string;
  salt_state_files?: string;
  target_groups?: string;
  target_hosts?: string;
  environment?: string;
  config_data?: string;
  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface Deployment {
  id: number;
  config_id: number;
  version: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
  started_by: number;
  started_at: string;
  completed_at?: string;
  duration_seconds?: number;
  target_hosts: string; // JSON array string
  salt_job_id?: string;
  results?: string; // JSON string
  error_message?: string;
  is_rollback: boolean;
  rollback_from_deployment_id?: number;
  created_at: string;
  updated_at: string;
  config?: DeploymentConfig;
}

export interface DeploymentVersion {
  id: number;
  application_name: string;
  version: string;
  release_notes?: string;
  created_by: number;
  created_at: string;
}

export interface ListDeploymentConfigsResponse {
  configs: DeploymentConfig[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListDeploymentsResponse {
  deployments: Deployment[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListDeploymentVersionsResponse {
  versions: DeploymentVersion[];
  total: number;
  page: number;
  page_size: number;
}

export interface CreateDeploymentConfigRequest {
  name: string;
  application_name: string;
  description?: string;
  salt_state_files?: string;
  target_groups?: string;
  target_hosts?: string;
  environment?: string;
  config_data?: string;
}

export interface CreateDeploymentRequest {
  config_id: number;
  version: string;
  target_hosts: number[];
}

export interface CreateDeploymentVersionRequest {
  application_name: string;
  version: string;
  release_notes?: string;
}

export const deploymentService = {
  // 部署配置
  listConfigs: async (
    page: number = 1,
    pageSize: number = 10,
    filters?: { application_name?: string; name?: string; environment?: string }
  ): Promise<ListDeploymentConfigsResponse> => {
    const response = await api.get('/deployment-configs', {
      params: { page, page_size: pageSize, ...filters },
    });
    return response.data;
  },

  getConfig: async (id: number): Promise<{ config: DeploymentConfig }> => {
    const response = await api.get(`/deployment-configs/${id}`);
    return response.data;
  },

  createConfig: async (data: CreateDeploymentConfigRequest): Promise<{ config: DeploymentConfig }> => {
    const response = await api.post('/deployment-configs', data);
    return response.data;
  },

  updateConfig: async (id: number, data: Partial<CreateDeploymentConfigRequest>): Promise<void> => {
    await api.put(`/deployment-configs/${id}`, data);
  },

  deleteConfig: async (id: number): Promise<void> => {
    await api.delete(`/deployment-configs/${id}`);
  },

  // 部署执行
  listDeployments: async (
    page: number = 1,
    pageSize: number = 10,
    filters?: { config_id?: number; status?: string; version?: string }
  ): Promise<ListDeploymentsResponse> => {
    const response = await api.get('/deployments', {
      params: { page, page_size: pageSize, ...filters },
    });
    return response.data;
  },

  getDeployment: async (id: number): Promise<{ deployment: Deployment }> => {
    const response = await api.get(`/deployments/${id}`);
    return response.data;
  },

  getDeploymentStatus: async (id: number): Promise<{ deployment: Deployment }> => {
    const response = await api.get(`/deployments/${id}/status`);
    return response.data;
  },

  createDeployment: async (data: CreateDeploymentRequest): Promise<{ deployment: Deployment }> => {
    const response = await api.post('/deployments', data);
    return response.data;
  },

  rollbackDeployment: async (id: number): Promise<{ deployment: Deployment }> => {
    const response = await api.post(`/deployments/${id}/rollback`);
    return response.data;
  },

  // 部署版本
  listVersions: async (
    page: number = 1,
    pageSize: number = 10,
    filters?: { application_name?: string }
  ): Promise<ListDeploymentVersionsResponse> => {
    const response = await api.get('/deployment-versions', {
      params: { page, page_size: pageSize, ...filters },
    });
    return response.data;
  },

  getVersion: async (id: number): Promise<{ version: DeploymentVersion }> => {
    const response = await api.get(`/deployment-versions/${id}`);
    return response.data;
  },

  createVersion: async (data: CreateDeploymentVersionRequest): Promise<{ version: DeploymentVersion }> => {
    const response = await api.post('/deployment-versions', data);
    return response.data;
  },

  getVersionsByApplication: async (applicationName: string): Promise<{ versions: DeploymentVersion[] }> => {
    const response = await api.get(`/deployment-versions/application/${applicationName}`);
    return response.data;
  },
};

File: backend/internal/models/deployment.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// DeploymentConfig 部署配置模型
type DeploymentConfig struct {
	ID             uint64         `gorm:"primaryKey" json:"id"`
	Name           string         `gorm:"size:100;not null;index" json:"name"`
	ApplicationName string        `gorm:"size:100;not null;index" json:"application_name"`
	Description    string         `json:"description"`
	SaltStateFiles string         `gorm:"type:text[]" json:"salt_state_files"` // PostgreSQL array
	TargetGroups   string         `gorm:"type:bigint[]" json:"target_groups"` // PostgreSQL array
	TargetHosts    string         `gorm:"type:bigint[]" json:"target_hosts"`   // PostgreSQL array
	Environment    string         `gorm:"size:50" json:"environment"`          // dev, test, prod
	ConfigData     string         `gorm:"type:jsonb" json:"config_data"`        // JSON格式
	CreatedBy      uint64         `gorm:"not null;index" json:"created_by"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	Creator *User        `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
	Deployments []Deployment `gorm:"foreignKey:ConfigID" json:"deployments,omitempty"`
}

// TableName 指定表名
func (DeploymentConfig) TableName() string {
	return "deployment_configs"
}

// Deployment 部署执行记录模型
type Deployment struct {
	ID                      uint64         `gorm:"primaryKey" json:"id"`
	ConfigID                uint64         `gorm:"not null;index" json:"config_id"`
	Version                 string         `gorm:"size:50;not null" json:"version"`
	Status                  string         `gorm:"size:20;not null;default:'pending';index" json:"status"` // pending, running, completed, failed, cancelled
	StartedBy               uint64         `gorm:"not null;index" json:"started_by"`
	StartedAt               time.Time      `gorm:"not null;default:now()" json:"started_at"`
	CompletedAt             *time.Time     `json:"completed_at"`
	DurationSeconds         *int           `json:"duration_seconds"`
	TargetHosts             string         `gorm:"type:bigint[];not null" json:"target_hosts"` // PostgreSQL array
	SaltJobID               string         `gorm:"size:255;index" json:"salt_job_id"`
	Results                 string         `gorm:"type:jsonb" json:"results"` // JSON格式，按主机存储结果
	ErrorMessage            string         `gorm:"type:text" json:"error_message"`
	IsRollback              bool           `gorm:"not null;default:false" json:"is_rollback"`
	RollbackFromDeploymentID *uint64       `gorm:"index" json:"rollback_from_deployment_id"`
	CreatedAt               time.Time      `json:"created_at"`
	UpdatedAt               time.Time      `json:"updated_at"`

	// 关联
	Config                  *DeploymentConfig `gorm:"foreignKey:ConfigID" json:"config,omitempty"`
	Starter                 *User             `gorm:"foreignKey:StartedBy" json:"starter,omitempty"`
	RollbackFromDeployment  *Deployment       `gorm:"foreignKey:RollbackFromDeploymentID" json:"rollback_from_deployment,omitempty"`
}

// TableName 指定表名
func (Deployment) TableName() string {
	return "deployments"
}

// DeploymentVersion 部署版本模型
type DeploymentVersion struct {
	ID             uint64    `gorm:"primaryKey" json:"id"`
	ApplicationName string   `gorm:"size:100;not null;index" json:"application_name"`
	Version        string    `gorm:"size:50;not null" json:"version"`
	ReleaseNotes   string    `json:"release_notes"`
	CreatedBy      uint64    `gorm:"not null;index" json:"created_by"`
	CreatedAt      time.Time `json:"created_at"`

	// 关联
	Creator *User `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
}

// TableName 指定表名
func (DeploymentVersion) TableName() string {
	return "deployment_versions"
}

